File: trendDetector.js
/* trendDetector.js — regime-change detection for the night card.
 *
 * Port of trend_detector.py. Same rule, same constants, same output.
 *
 * The card's fallthrough is unconditional: when nothing crosses a band it has
 * to say something, and "probably just one night" is the only thing left. That
 * means a developmental step gets narrated as noise every night it continues.
 * Call verdict() before the fallthrough and use its text when kind === 'trend'.
 *
 * Input: nights sorted ascending, each
 *   { date: 'YYYY-MM-DD', complete: bool, totalH, longestH, daySleepH,
 *     blocks: int, onsetMin: minutes after 18:00 }
 */

export const RECENT = 7;
export const PRIOR = 14;
export const MIN_SD_GAP = 1.5;
export const MIN_COUNT = 4;
const MIN_HISTORY = RECENT + PRIOR;

// key -> [sd floor in metric units, rising label, falling label]
export const METRICS = {
  longestH:  [0.50, 'consolidating',  'fragmenting'],
  totalH:    [0.35, 'lengthening',    'shortening'],
  daySleepH: [0.45, 'napping more',   'napping less'],
  blocks:    [0.60, 'waking more',    'waking less'],
  onsetMin:  [20.0, 'settling later', 'settling earlier'],
};

const mean = a => a.reduce((s, v) => s + v, 0) / a.length;
const pstdev = a => {
  const m = mean(a);
  return Math.sqrt(mean(a.map(v => (v - m) ** 2)));
};

function history(nights, targetIndex, key) {
  return nights
    .slice(0, targetIndex + 1)
    .filter(n => n.complete && Number.isFinite(n[key]))
    .map(n => n[key]);
}

export function detect(nights, targetIndex, key) {
  const values = history(nights, targetIndex, key);
  if (values.length < MIN_HISTORY) return null;

  const recent = values.slice(-RECENT);
  const prior = values.slice(-(RECENT + PRIOR), -RECENT);
  const [sdFloor, rising, falling] = METRICS[key];

  const mu = mean(prior);
  const sd = Math.max(pstdev(prior), sdFloor);
  const sdGap = (mean(recent) - mu) / sd;
  if (Math.abs(sdGap) < MIN_SD_GAP) return null;

  const up = sdGap > 0;
  const bound = up ? mu + sd : mu - sd;
  const outside = recent.filter(v => (up ? v > bound : v < bound)).length;
  if (outside < MIN_COUNT) return null;

  return {
    metric: key,
    label: up ? rising : falling,
    direction: up ? 'up' : 'down',
    sdGap,
    priorMean: mu,
    recentMean: mean(recent),
    outside,
  };
}

/* Walk back to the first night this trend fired, so the card can say how long
 * it has been running instead of re-announcing it every night. */
export function firstFired(nights, targetIndex, key) {
  const current = detect(nights, targetIndex, key);
  if (!current) return targetIndex;
  let earliest = targetIndex;
  for (let i = targetIndex - 1; i >= 0; i--) {
    if (!nights[i].complete) continue;
    const past = detect(nights, i, key);
    if (!past || past.direction !== current.direction) break;
    earliest = i;
  }
  return earliest;
}

export function detectAll(nights, targetIndex) {
  const hits = [];
  for (const key of Object.keys(METRICS)) {
    const hit = detect(nights, targetIndex, key);
    if (!hit) continue;
    const since = firstFired(nights, targetIndex, key);
    hit.since = nights[since].date;
    hit.nightsRunning =
      Math.round(
        (Date.parse(nights[targetIndex].date) - Date.parse(hit.since)) / 864e5
      ) + 1;
    hits.push(hit);
  }
  hits.sort((a, b) => Math.abs(b.sdGap) - Math.abs(a.sdGap));
  return hits;
}

const hm = h => `${Math.floor(h)}h${String(Math.round((h % 1) * 60)).padStart(2, '0')}m`;
const clock = m => {
  const t = 18 * 60 + m;
  return `${String(Math.floor(t / 60) % 24).padStart(2, '0')}:${String(Math.round(t % 60)).padStart(2, '0')}`;
};

export function narrate(t) {
  const run = t.nightsRunning;
  const tail = run > 1
    ? `Holding for ${run} nights now.`
    : 'First night this reads as a trend.';

  let head, body;
  switch (t.metric) {
    case 'longestH':
      head = t.direction === 'up'
        ? 'His longest stretch is consolidating.'
        : 'His longest stretch is breaking up.';
      body = `It has averaged ${hm(t.recentMean)} over the last ${RECENT} nights against `
           + `${hm(t.priorMean)} for the ${PRIOR} before, with ${t.outside} of ${RECENT} `
           + 'nights outside the old range.';
      break;
    case 'totalH':
      head = t.direction === 'up'
        ? 'He is sleeping longer at night overall.'
        : 'Night sleep is trending down.';
      body = `${hm(t.recentMean)} on average over ${RECENT} nights against ${hm(t.priorMean)} before.`;
      break;
    case 'daySleepH':
      head = t.direction === 'up'
        ? 'Day sleep is climbing.'
        : 'Day sleep is dropping off.';
      body = `${hm(t.recentMean)} a day over the last ${RECENT} against ${hm(t.priorMean)} before, `
           + `${t.outside} of ${RECENT} days outside the old range.`;
      break;
    case 'blocks':
      head = t.direction === 'up' ? 'He is waking more often.' : 'He is waking less often.';
      body = `${t.recentMean.toFixed(1)} blocks a night against ${t.priorMean.toFixed(1)} before.`;
      break;
    default:
      head = t.direction === 'up' ? 'Bedtime is drifting later.' : 'Bedtime is moving earlier.';
      body = `Onset now averages ${clock(t.recentMean)} against ${clock(t.priorMean)} before.`;
  }
  return `${head} ${body} ${tail}`;
}

/* Ranks which metric to lead with when no trend is running.
 * Not by raw variability — the noisiest series is the one where a single day
 * outside its band means least. Rank by how far the value sits from its own
 * band in units of that metric's own spread. */
export function rankByExcursion(nights, targetIndex) {
  const night = nights[targetIndex];
  return Object.keys(METRICS)
    .map(key => {
      const values = history(nights, targetIndex - 1, key);
      if (values.length < 5 || !Number.isFinite(night[key])) return null;
      const window = values.slice(-21);
      const sd = Math.max(pstdev(window), METRICS[key][0]);
      return { metric: key, z: (night[key] - mean(window)) / sd };
    })
    .filter(Boolean)
    .sort((a, b) => Math.abs(b.z) - Math.abs(a.z));
}

export function verdict(nights, targetIndex) {
  const night = nights[targetIndex];
  if (!night.complete) {
    return { kind: 'incomplete', text: 'Open sleep record with no bounding event. No number reported.' };
  }

  const trends = detectAll(nights, targetIndex);
  if (trends.length) {
    return { kind: 'trend', text: trends.map(narrate).join(' '), trends };
  }

  const ranked = rankByExcursion(nights, targetIndex);
  const lead = ranked[0];
  if (!lead || Math.abs(lead.z) < 1) {
    return { kind: 'in range', text: 'Inside his usual range on everything that matters.' };
  }
  return {
    kind: 'single night',
    text: 'Probably just one night. No trend behind it yet, so worth a second look tomorrow '
        + 'rather than a change tonight.',
    ranked,
  };
}

File: trend_detector.py
"""
Trend detector for the "Is this normal for your baby?" card.

Sits on top of night_reference.assemble_nights(). The shipped card compares one
night against a pooled band and, when nothing crosses it, falls through to
"probably just one night". That fallthrough is unconditional, so a regime change
can never surface. This module runs first and, when a trend is active, replaces
the single-night verdict with a statement about the trend.

Detection rule per metric:
    - trailing RECENT nights vs the PRIOR nights before them
    - fire when |mean gap| >= MIN_SD_GAP * sd(prior)  AND
      at least MIN_COUNT of the recent nights fall outside the prior band
    - sd floored at SD_FLOOR[metric] so a freakishly stable stretch cannot
      make a trivial shift look enormous
    - incomplete nights excluded from both windows

Both conditions matter. The sd gap alone fires on a single extreme night; the
count alone fires on a slow drift with no magnitude.
"""

import datetime as dt
import statistics as stats

import night_reference as nr

RECENT = 7
PRIOR = 14
MIN_SD_GAP = 1.5
MIN_COUNT = 4
MIN_HISTORY = RECENT + PRIOR

# metric -> (accessor, unit, sd floor in metric units, rising label, falling label)
METRICS = {
    "longest_h": (lambda n: n["longest_h"], "h", 0.50,
                  "consolidating", "fragmenting"),
    "total_h": (lambda n: n["total_h"], "h", 0.35,
                "lengthening", "shortening"),
    "blocks": (lambda n: float(n["blocks"]), "blocks", 0.60,
               "waking more", "waking less"),
    "onset_min": (lambda n: (n["onset"] - n["onset"].replace(
        hour=18, minute=0, second=0)).total_seconds() / 60, "min", 20.0,
        "settling later", "settling earlier"),
}


def series(nights, target, metric):
    keys = [d for d in sorted(nights) if d <= target and nights[d]["complete"]]
    values = [METRICS[metric][0](nights[d]) for d in keys]
    return keys, values


def detect(nights, target, metric):
    keys, values = series(nights, target, metric)
    if len(values) < MIN_HISTORY:
        return None

    recent = values[-RECENT:]
    prior = values[-(RECENT + PRIOR):-RECENT]
    _, unit, sd_floor, rising, falling = METRICS[metric]

    mu = stats.mean(prior)
    sd = max(stats.pstdev(prior), sd_floor)
    gap = (stats.mean(recent) - mu) / sd
    if abs(gap) < MIN_SD_GAP:
        return None

    bound = mu + sd if gap > 0 else mu - sd
    outside = sum(1 for v in recent if (v > bound if gap > 0 else v < bound))
    if outside < MIN_COUNT:
        return None

    return {
        "metric": metric,
        "label": rising if gap > 0 else falling,
        "direction": "up" if gap > 0 else "down",
        "sd_gap": gap,
        "prior_mean": mu,
        "recent_mean": stats.mean(recent),
        "prior_bound": bound,
        "outside": outside,
        "unit": unit,
    }


def first_fired(nights, target, metric):
    """Walk back to the first night this trend fired, so the card can say how
    long it has been running instead of re-announcing it nightly."""
    keys = [d for d in sorted(nights) if d <= target and nights[d]["complete"]]
    current = detect(nights, target, metric)
    if not current:
        return None
    earliest = target
    for d in reversed(keys[:-1]):
        past = detect(nights, d, metric)
        if not past or past["direction"] != current["direction"]:
            break
        earliest = d
    return earliest


def detect_all(nights, target):
    found = []
    for metric in METRICS:
        hit = detect(nights, target, metric)
        if hit:
            hit["since"] = first_fired(nights, target, metric)
            hit["nights_running"] = (target - hit["since"]).days + 1
            found.append(hit)
    found.sort(key=lambda h: -abs(h["sd_gap"]))
    return found


def fmt(value, unit):
    if unit == "h":
        return nr.hm(value)
    if unit == "min":
        base = dt.datetime(2000, 1, 1, 18, 0) + dt.timedelta(minutes=value)
        return base.strftime("%H:%M")
    return f"{value:.1f}"


def narrate(trend, nights, target):
    m, unit = trend["metric"], trend["unit"]
    prior, recent = fmt(trend["prior_mean"], unit), fmt(trend["recent_mean"], unit)
    run = trend["nights_running"]
    tail = (f"Holding for {run} nights now." if run > 1 else "First night this reads as a trend.")

    if m == "longest_h":
        head = ("His longest stretch is consolidating." if trend["direction"] == "up"
                else "His longest stretch is breaking up.")
        body = (f"It has averaged {recent} over the last {RECENT} nights against "
                f"{prior} for the {PRIOR} before, with {trend['outside']} of {RECENT} "
                f"nights outside the old range.")
    elif m == "total_h":
        head = ("He is sleeping longer at night overall." if trend["direction"] == "up"
                else "Night sleep is trending down.")
        body = f"{recent} on average over {RECENT} nights against {prior} before."
    elif m == "blocks":
        head = ("He is waking more often." if trend["direction"] == "up"
                else "He is waking less often.")
        body = (f"{trend['recent_mean']:.1f} blocks a night against "
                f"{trend['prior_mean']:.1f} before.")
    else:
        head = ("Bedtime is drifting later." if trend["direction"] == "up"
                else "Bedtime is moving earlier.")
        body = f"Onset now averages {recent} against {prior} before."

    return f"{head} {body} {tail}"


def verdict(nights, target):
    """Trend first; single-night framing only when nothing is running."""
    night = nights[target]
    if not night["complete"]:
        return "incomplete", "Open sleep record with no bounding event. No number reported."

    trends = detect_all(nights, target)
    if trends:
        lead = narrate(trends[0], nights, target)
        rest = [narrate(t, nights, target) for t in trends[1:]]
        return "trend", " ".join([lead] + rest)

    base = nr.evaluate(nights, target)
    if base["total_where"] == "inside" and base["longest_where"] == "inside":
        return "in range", "Inside his usual range on both total and longest block."
    return "single night", ("Outside range on "
                            + ", ".join(k for k, v in
                                        (("total", base["total_where"]),
                                         ("longest block", base["longest_where"]))
                                        if v != "inside")
                            + ". No trend behind it yet, so worth a second look tomorrow "
                              "rather than a change tonight.")


if __name__ == "__main__":
    import sys
    path = sys.argv[1]
    nights = nr.assemble_nights(nr.close_open_sleeps(nr.load_events(path)))
    keys = [d for d in sorted(nights) if nights[d]["complete"]]

    if "--backtest" in sys.argv:
        print("firing history (every night with enough history):\n")
        for d in keys[MIN_HISTORY:]:
            kind, text = verdict(nights, d)
            if kind == "trend":
                print(f"{d}  {text}")
        print("\nfiring rate:")
        fires = sum(1 for d in keys[MIN_HISTORY:] if verdict(nights, d)[0] == "trend")
        n = len(keys) - MIN_HISTORY
        print(f"  {fires} of {n} nights ({100 * fires / n:.0f}%)")
    else:
        target = keys[-1]
        kind, text = verdict(nights, target)
        print(f"night of {target} -> {kind}\n{text}")
